File: scala/src/test/scala/anti/patterns/BlogTests.scala
package anti.patterns

import demo.blog.{Article, BlogService, Comment}
import org.scalatest.EitherValues
import org.scalatest.flatspec.AnyFlatSpec

import java.time.LocalDate

class BlogTests extends AnyFlatSpec with EitherValues {

  it should "returns a Right for valid comment" in {
    val blogService = new BlogService()
    val article = Article(
      "Lorem Ipsum",
      "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
    )
    val result = blogService.addComment(article, "Amazing article !!!", "Pablo Escobar")

    assert(result.isRight)
  }

  it should "add a comment with the given text" in {
    val blogService = new BlogService()
    val article = Article(
      "Lorem Ipsum",
      "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
    )
    val text = "Amazing article !!!"
    val result = blogService.addComment(article, text, "Pablo Escobar")

    assert(result.value.comments.head.text == text)
  }

  it should "add a comment with the given author" in {
    val blogService = new BlogService()
    val article = Article(
      "Lorem Ipsum",
      "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
    )
    val author = "Pablo Escobar"
    val result = blogService.addComment(article, "Amazing article !!!", author)

    assert(result.value.comments.head.author == author)
  }

  it should "add a comment with the date of the day" in {
    val blogService = new BlogService()
    val article = Article(
      "Lorem Ipsum",
      "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
    )
    val author = "Pablo Escobar"
    val result = blogService.addComment(article, "Amazing article !!!", author)
  }

  it should "returns a Left when adding existing comment" in {
    val blogService = new BlogService()
    val article = Article(
      "Lorem Ipsum",
      "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore",
      List(Comment("Amazing article !!!", "Pablo Escobar", LocalDate.now()))
    )
    val result = blogService
      .addComment(article, "Amazing article !!!", "Pablo Escobar")

    assert(result.isLeft)
  }
}

File: php/tests/AntiPatterns/BlogTests.php
<?php

namespace AntiPatterns;

use Com\CleanTests\Blog\Article;
use PHPUnit\Framework\TestCase;
use function PHPUnit\Framework\assertCount;
use function PHPUnit\Framework\assertEquals;

class BlogTests extends TestCase
{
    public function testShouldReturnNothingForAValidComment(): void
    {
        $article = new Article(
            "Lorem Ipsum",
            "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );

        $result = $article->addComment("Amazing article !!!", "Pablo Escobar");

        assertCount(0, $result);
    }

    public function testShouldAddACommentWithTheGivenText(): void
    {
        $article = new Article(
            "Lorem Ipsum",
            "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );

        $text = "Amazing article !!!";
        $article->addComment($text, "Pablo Escobar");

        assertCount(1, $article->getComments());
        assertEquals($text, $article->getComments()[0]->getText());
    }

    public function testShouldAddACommentWithTheGivenAuthor(): void
    {
        $article = new Article(
            "Lorem Ipsum",
            "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );

        $author = "Pablo Escobar";
        $article->addComment("Amazing article !!!", $author);

        assertCount(1, $article->getComments());
        assertEquals($author, $article->getComments()[0]->getAuthor());
    }

    public function testShouldAddACommentWithTheDateOfTheDay(): void
    {
        $article = new Article(
            "Lorem Ipsum",
            "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );

        $article->addComment("Amazing article !!!", "Pablo Escobar");
        assertCount(1, $article->getComments());
    }

    public function testShouldFailWhenAddingExistingComment(): void
    {
        $article = new Article(
            "Lorem Ipsum",
            "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );

        $article->addComment("Amazing article !!!", "Pablo Escobar");
        $result = $article->addComment("Amazing article !!!", "Pablo Escobar");

        assertCount(1, $article->getComments());
        assertCount(1, $result);
        assertEquals("Comment already in the article", $result[0]->getDescription());
    }
}